commands/completion.go

```diff
@@ -0,0 +1,223 @@
+package commands
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+)
+
+var shell string
+
+var completionCmd = &cobra.Command{
+	Use:   "completion SHELL",
+	Short: "Generates shell auto completion",
+	Long:  "Generates shell auto completion for Bash or ZSH.",
+	Example: `  faas-cli completion --shell bash
+  faas-cli completion --shell zsh`,
+	RunE: runCompletion,
+}
+
+func init() {
+	completionCmd.Flags().StringVar(&shell, "shell", "", "Outputs shell completion, must be bash or zsh")
+	completionCmd.MarkFlagRequired("shell")
+
+	faasCmd.AddCommand(completionCmd)
+}
+
+func runCompletion(cmd *cobra.Command, args []string) (err error) {
+	if shell == "" {
+		return fmt.Errorf("--shell is required and must be bash or zsh")
+	}
+
+	switch shell {
+	case "bash":
+		err = generateBashCompletion()
+		if err != nil {
+			return err
+		}
+		return nil
+
+	case "zsh":
+		err = generateZshCompletion()
+		if err != nil {
+			return err
+		}
+		return nil
+
+	default:
+		return fmt.Errorf("%q shell not supported, must be bash or zsh", shell)
+	}
+}
+
+func generateBashCompletion() error {
+	err := faasCmd.GenBashCompletion(os.Stdout)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func generateZshCompletion() error {
+	zshHead := "#compdef faas-cli\n"
+
+	out := os.Stdout
+
+	_, err := out.Write([]byte(zshHead))
+	if err != nil {
+		return err
+	}
+
+	zshInitialization := `
+__faas-cli_bash_source() {
```

Comment:
This seems like a lot of text to have embedded in a quoted string. Are there any alternatives? How do other Golang tools do this like hashicorp or docker?